### ISSUE TITLE
save -r doesn't save inside subdataset when called with -d^, requires second call

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#3943]
Minimal example. You can see that on 2nd run of `save` it would add that file.
<details>
<summary>longer example - but apparently superds is not needed to exhibit bug</summary>
```
/tmp > datalad create superds && cd superds && datalad create -d . testds && cd testds && mkdir -p subds1 && touch subds1/file && datalad create -f -d subds1
[INFO   ] Creating a new annex repo at /tmp/superds 
create(ok): /tmp/superds (dataset)                                                                                               
[INFO   ] Creating a new annex repo at /tmp/superds/testds 
add(ok): testds (file)                                                                                                           
add(ok): .gitmodules (file)
save(ok): . (dataset)
create(ok): testds (dataset)
action summary:
  add (ok: 2)
  create (ok: 1)
  save (ok: 1)
[INFO   ] Creating a new annex repo at /tmp/superds/testds/subds1 
create(ok): . (dataset)                                                                                                          

/tmp/superds/testds > datalad save -r -d^ subds1
add(ok): subds1 (file)
add(ok): .gitmodules (file)
save(ok): testds (dataset)
add(ok): testds (file)
save(ok): . (dataset)
action summary:
  add (ok: 3)
  save (ok: 2)

/tmp/superds/testds > datalad save -r -d^ subds1
add(ok): file (file)
save(ok): testds/subds1 (dataset)
add(ok): subds1 (file)
save(ok): testds (dataset)
add(ok): testds (file)
save(ok): . (dataset)
action summary:
  add (ok: 3)
  save (ok: 3)

/tmp/superds/testds > datalad save -r -d^ subds1
action summary:
  save (notneeded: 3)

/tmp/superds/testds > datalad --version
datalad 0.12.0rc6.dev198
```
</details>

```shell
exit:1 /tmp > datalad create testds && cd testds && mkdir -p subds1 && touch subds1/file && datalad create -f -d subds1 
[INFO   ] Creating a new annex repo at /tmp/testds 
create(ok): /tmp/testds (dataset)                                                                                                
[INFO   ] Creating a new annex repo at /tmp/testds/subds1 
create(ok): . (dataset)                                                                                                          

/tmp/testds > datalad save -r -d^ subds1                                                                               
add(ok): subds1 (file)
add(ok): .gitmodules (file)
save(ok): . (dataset)
action summary:
  add (ok: 2)
  save (ok: 1)

/tmp/testds > datalad save -r -d^ subds1
add(ok): file (file)
save(ok): subds1 (dataset)
add(ok): subds1 (file)
save(ok): . (dataset)
action summary:
  add (ok: 2)
  save (ok: 2)
```
and it is the `-d` specified which affects it (`-d.` is as "good"):
```shell
/tmp > datalad create testds && cd testds && mkdir -p subds1 && touch subds1/file && datalad create -f -d subds1
[INFO   ] Creating a new annex repo at /tmp/testds 
create(ok): /tmp/testds (dataset)                                                                                                
[INFO   ] Creating a new annex repo at /tmp/testds/subds1 
create(ok): . (dataset)                                                                                                          

/tmp/testds > datalad save -d. -r subds1                                                                               
add(ok): subds1 (file)
add(ok): .gitmodules (file)
save(ok): . (dataset)
action summary:
  add (ok: 2)
  save (ok: 1)

/tmp/testds > datalad save -d. -r subds1
add(ok): file (file)
save(ok): subds1 (dataset)
add(ok): subds1 (file)
save(ok): . (dataset)
action summary:
  add (ok: 2)
  save (ok: 2)
```

**Comment by yarikoptic** [comment]
I guess it is because upon initial invocation of `save`, `subds` is not yet a known to it subdataset, so `-r` is not in effect.  But I would have expected that it would save (probably first within since it is to be committed, but could be first saved to be added, and then saved within, and then saved new state on top).
The same happens even without pointing to it with `save -r .` which should in theory save everything:
```
/tmp/testds > datalad save -r                                                                                          
add(ok): subds1 (file)
add(ok): .gitmodules (file)
save(ok): . (dataset)
action summary:
  add (ok: 2)
  save (ok: 1)

/tmp/testds > datalad save -r
add(ok): file (file)
save(ok): subds1 (dataset)
add(ok): subds1 (file)
save(ok): . (dataset)
action summary:
  add (ok: 2)
  save (ok: 2)
```

**Comment by yarikoptic** [comment]
0.11.x "analogue" `add` works out this case correctly
```
$> datalad add -d^ -r subds1                                                                               
add(ok): subds1/file (file)
add(ok): subds1 (dataset)
add(ok): subds1 (dataset) [added new subdataset]
save(ok): /tmp/testds/subds1 (dataset)
save(ok): /tmp/testds (dataset)
action summary:
  add (notneeded: 2, ok: 3)
  save (ok: 2)

$> datalad add -d^ -r subds1
$>
```

**Comment by mih** [comment]
IIRC that is an intentional change.

**Comment by yarikoptic** [comment]
mnah, comparing `datalad save` to `publish` here is like comparing `git add` to `push`. Sure thing `push` shouldn't (and will not) `push` unregistered files neither on the first, nor on the second (re)invocation. But the sole role of `save` (as a replacement of `add` and the ultimate ideologist "from clean state to clean state" tracking) is to save all the changes to get to a clean state in a single invocation and not a chain of a sufficiently long identical `save`s calls until it finally arrives to a clean state. And `-r` here is the (only) way to say for everything, including submodules. 
So, if there was any intention, it feels to be more implementation- not a usecase- (or "clean semantic") driven.  Unless there was actually some use case such behavior helped to address?

**Comment by mih** [comment]
I don't think it is useful to employ git analogies to figure out what datalad should do. We are talking about functionality that git doesn't have AFAIK (`git submodule foreach` doesn't seem to touch non-submodule repositories.

I take what you are saying as the proposal to overload `--recursive` (generally being recursion into subdatasets, because directory recursion within datasets is typically implied and generally unavoidable), with also being recursion into subdirectories, when it could make sense, but needs to be enabled, although typically implied.

So when is directory-recursion into non-subdatasets OK when uniformity for `--recursion` is not OK?

- [ ] aggregate-metadata
- [ ] diff
- [ ] drop
- [ ] get
- [ ] install
- [ ] ls (already advertises recursion as "subdirectory recursion")
- [ ] metadata
- [ ] publish
- [ ] remove
- [ ] save
- [ ] siblings
- [ ] status
- [ ] subdatasets
- [ ] uninstall
- [ ] unlock,
- [ ] update

**Comment by yarikoptic** [comment]
so far recursion into subdirectories is needed only for `save` (per original point of this issue).  So far I do not think is pertinent to any other command (but if I mark them in your checklist - we wouldn't know if it was you or not ;))
May be `ls` but that one is an antique alien, meaning of `-r` in which ATM depends on what it is provided with -- for datasets - would explore subdatasets, for directories and files - does not recurse at all (from top level interface at least), for s3 urls - would recurse into "directories" (there is no directories per se on s3).

**Comment by mih** [comment]
What should a `--recursion-limit` limit do in case of `--recursive` meaning recursion into subdirectories?

**Comment by yarikoptic** [comment]
nothing.  my view: `--recursion` (and its limit) is about subdatasets and (at least ATM) it should stay that way. I do not think it would be productive to dive into all "subdatasets vs subdirectories" semantics/discussion here.
Even now `save -r` alters the listing of subdatasets, but does not care to apply `save` to the subdataset it just added (2nd invocation does it).  So IMHO a solution to this particular issue could be  for `save` internally trigger `save`'ing freshly added (by it) subdatasets (if ran with `--recursive` and respecting the limit) -- so it would not solely rely on listing of `subdatasets` as when initially invoked.  The simplest solution would generate two commits in the superdataset -- one for adding subdataset (as is), then 2nd one (possibly, if subdataset was dirty) is to save updated state on the dataset after it calls `save -r` within that dataset (adjusting recursion limit how it already does).

**Comment by yarikoptic** [comment]
BTW, just spotted in

> ```shell
> /tmp/testds > datalad save -r -d^ subds1                                                                               
> add(ok): subds1 (file)
> add(ok): .gitmodules (file)
> ...
> ```

is incorrectly lists `subds1` as a `(file)`. It is not, either in human or git (where it would be "Subproject" or smth like that IIRC)  or datalad (it is a dataset already, just was not a sub-dataset) terms.  Ideally should be

> ```shell
> /tmp/testds > datalad save -r -d^ subds1                                                                               
> add(ok): .gitmodules (file)
> add(ok): subds1 (dataset)
> ...
> ```

(note that 0.11.x was reporting it as a dataset as well). I also reordered (not sure if easy/feasible) to make it even more kosher, i.e. that by the time `add` reported on `subds1`, `.gitmodules` also add'ed.  But that is minor and could be ignored -- it doesn't change the nature of subds1 by itself being a dataset

**Comment by yarikoptic** [comment]
the issue remains, an additional point to mention that current behavior is not inline with the documentation of `save --help`:
```rst
Save any content underneath the current directory, and
recurse into any potential subdatasets::

   % datalad save . --recursive
```
given that `save` does add those potential subdatasets in the first invocation, so it does treat them as subdatasets, it should also save their state.

**Comment by mih** [comment]
FTR: I do not plan to work on this, my views have not changed. I can provide a documentation fix to say:

```
Save any files underneath the current directory to the dataset, and
recurse into any potential subdatasets::
```
